modules/HashProtocol.js

```diff
@@ -61,9 +61,13 @@ export const startListener = (listener, pathCoder, queryKey) => {
       replaceHashPath(encodedPath)
     } else {
       const currentLocation = getCurrentLocation(pathCoder, queryKey)
+      const isEqualKey = prevLocation && currentLocation.key && prevLocation.key === currentLocation.key
+      const isEqualPathname = prevLocation && prevLocation.pathname === currentLocation.pathname
+      const isEqualSearch = prevLocation && prevLocation.search === currentLocation.search
 
-      if (prevLocation && currentLocation.key && prevLocation.key === currentLocation.key)
+      if (isEqualKey || isEqualPathname && isEqualSearch) {
```

Comment:
This logic isn't quite right. See perhaps https://github.com/ReactTraining/history/blob/f247b4001cd1ad8e477a1e131cada1cba58fc1fe/modules/createHistory.js#L100-L106.